pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go

```diff
@@ -1915,283 +1939,304 @@ func validateVolumes(field *k8sfield.Path, volumes []v1.Volume, config *virtconf
-		if volume.PersistentVolumeClaim != nil {
-			volumeSourceSetCount++
-		}
-		if volume.Sysprep != nil {
-			volumeSourceSetCount++
-		}
-		if volume.CloudInitNoCloud != nil {
-			volumeSourceSetCount++
-		}
-		if volume.CloudInitConfigDrive != nil {
-			volumeSourceSetCount++
-		}
-		if volume.ContainerDisk != nil {
-			volumeSourceSetCount++
-		}
-		if volume.Ephemeral != nil {
-			volumeSourceSetCount++
-		}
-		if volume.EmptyDisk != nil {
-			volumeSourceSetCount++
-		}
-		if volume.HostDisk != nil {
-			volumeSourceSetCount++
-		}
-		if volume.DataVolume != nil {
-			if !config.HasDataVolumeAPI() {
-				causes = append(causes, metav1.StatusCause{
-					Type:    metav1.CauseTypeFieldValueInvalid,
-					Message: "DataVolume api is not present in cluster. CDI must be installed for DataVolume support.",
-					Field:   field.Index(idx).String(),
-				})
-			}
+		causes = append(causes, validateVolume(field.Index(idx), volume, config)...)
+	}
 
-			if volume.DataVolume.Name == "" {
-				causes = append(causes, metav1.StatusCause{
-					Type:    metav1.CauseTypeFieldValueRequired,
-					Message: "DataVolume 'name' must be set",
-					Field:   field.Index(idx).Child("name").String(),
-				})
-			}
-			volumeSourceSetCount++
-		}
-		if volume.ConfigMap != nil {
-			volumeSourceSetCount++
-		}
-		if volume.Secret != nil {
-			volumeSourceSetCount++
-		}
-		if volume.DownwardAPI != nil {
-			volumeSourceSetCount++
-		}
-		if volume.ServiceAccount != nil {
-			volumeSourceSetCount++
+	// check that we have max 1 instance of below disks
+	serviceAccountVolumeCount := 0
+	downwardMetricVolumeCount := 0
+	memoryDumpVolumeCount := 0
+
+	for i := range volumes {
+		if volumes[i].ServiceAccount != nil {
 			serviceAccountVolumeCount++
 		}
-		if volume.DownwardMetrics != nil {
+		if volumes[i].DownwardMetrics != nil {
 			downwardMetricVolumeCount++
-			volumeSourceSetCount++
 		}
-		if volume.MemoryDump != nil {
+		if volumes[i].MemoryDump != nil {
 			memoryDumpVolumeCount++
-			volumeSourceSetCount++
 		}
+	}
 
-		if volumeSourceSetCount != 1 {
+	if serviceAccountVolumeCount > 1 {
+		causes = append(causes, metav1.StatusCause{
+			Type:    metav1.CauseTypeFieldValueInvalid,
+			Message: fmt.Sprintf("%s must have max one serviceAccount volume set", field.String()),
+			Field:   field.String(),
+		})
+	}
+
+	if downwardMetricVolumeCount > 1 {
+		causes = append(causes, metav1.StatusCause{
+			Type:    metav1.CauseTypeFieldValueInvalid,
+			Message: fmt.Sprintf("%s must have max one downwardMetric volume set", field.String()),
+			Field:   field.String(),
+		})
+	}
+	if memoryDumpVolumeCount > 1 {
+		causes = append(causes, metav1.StatusCause{
+			Type:    metav1.CauseTypeFieldValueInvalid,
+			Message: fmt.Sprintf("%s must have max one memory dump volume set", field.String()),
+			Field:   field.String(),
+		})
+	}
+
+	return causes
+}
+
+func validateVolume(field *k8sfield.Path, volume v1.Volume, config *virtconfig.ClusterConfig) []metav1.StatusCause {
```

Comment:
This is a huge function. Can we split it to shorter functions?

Reply:
Done